charms/kfp-persistence/tests/unit/test_operator.py

```diff
@@ -1,262 +1,165 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-from contextlib import nullcontext as does_not_raise
+from typing import Optional
+from unittest.mock import MagicMock
 
-import ops
 import pytest
 import yaml
-from oci_image import MissingResourceError
 from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
 from ops.testing import Harness
 
-from charm import CheckFailedError, KfpPersistenceOperator
+from charm import KfpPersistenceOperator
 
-# TODO: Tests missing for dropped/reloaded relations
+MOCK_KFP_API_DATA = {"service-name": "service-name", "service-port": "1234"}
 
-ops.testing.SIMULATE_CAN_CONNECT = True
 
+@pytest.fixture
+def harness():
+    """Initialize Harness instance."""
+    harness = Harness(KfpPersistenceOperator)
+    harness.set_can_connect("persistenceagent", True)
+    return harness
 
-def test_not_leader(harness):
+
+def test_not_leader(
+    harness,  # noqa F811
+):
+    """Test not a leader scenario."""
     harness.begin_with_initial_hooks()
-    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    assert harness.charm.model.unit.status == WaitingStatus(
+        "[leadership-gate] Waiting for leadership"
+    )
 
 
-def test_image_fetch(harness, oci_resource_data):
+RELATION_NAME = "kfp-api"
+OTHER_APP_NAME = "kfp-api-provider"
+
+
+def test_kfp_api_relation_with_data(harness):
+    """Test that if Leadership is Active, the kfp-api relation operates as expected."""
+    # Arrange
     harness.begin()
-    with pytest.raises(MissingResourceError):
-        harness.charm.image.fetch()
 
-    harness.add_oci_resource(**oci_resource_data)
-    with does_not_raise():
-        harness.charm.image.fetch()
+    # Mock:
+    # * leadership_gate_component_item to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
 
+    # Add relation with data.  This should trigger a charm reconciliation due to relation-changed.
+    add_sdi_relation_to_harness(harness, "kfp-api", data=MOCK_KFP_API_DATA)
```

Comment:
This is now available in chisme using `from charmed_kubeflow_chisme.testing import add_sdi_relation_to_harness`